### PR TITLE
Test prepare options collision

### PR DESCRIPTION
**Pull request by jduss4**
With all the other options, we make it so that if the app setting was "num => 20" and you send in "num => 50", the request will override the app.

on the other hand, some fields like f[] and facet[] are arrays.
app might have something like f[]=collection|whitman set
and if a request sends in f[]=category|writings we don't want
it to completely override the existing array, nor do
we want it to blindly be added to the array if the request
is altering the collection filter
instead, selectively combine the arrays by filtering out any
app wide settings that would conflict with a newly requested one

### CODE REVIEW

File: test/api_bridge_test.rb [hunk 1]
```diff
@@ -27,7 +27,7 @@ def before_setup
     facets = ["creator.name", "date.year", "subcategory"]
     assert_equal facets, @api.app_facets
     assert_equal(
-      { "num" => 5, "sort" => "title|asc", "facet" => facets },
+      {"num"=>5, "sort"=>"title|asc", "f"=>["collection|letters"], "facet"=>["creator.name", "date.year", "subcategory"]},
```
Comment (techgique) [inline test/api_bridge_test.rb]: Would you mind restoring the whitespace? :pray: It looks like a really long line that should be split across lines too. Style guide conventions

File: test/api_bridge_test.rb [hunk 2]
```diff
@@ -122,7 +122,7 @@ def before_setup
     req_opts = ActionController::Parameters.new({
       controller: "items", action: "index", utf8: "✓", q: "water"
     })
-    res_opts = { "num"=>5, "sort"=>"title|asc", "facet"=>["creator.name", "date.year", "subcategory"], "q"=>"water", "start"=>0 }
+    res_opts = {"num"=>5, "sort"=>"title|asc", "f"=>["collection|letters"], "facet"=>["creator.name", "date.year", "subcategory"], "q"=>"water", "start"=>0}
```
Comment (techgique) [inline test/api_bridge_test.rb]: Whitespace and split lines again please? :)

File: test/api_bridge_test.rb [hunk 3]
```diff
@@ -27,7 +29,7 @@ def before_setup
     facets = ["creator.name", "date.year", "subcategory"]
     assert_equal facets, @api.app_facets
     assert_equal(
-      {"num"=>5, "sort"=>"title|asc", "f"=>["collection|letters"], "facet"=>["creator.name", "date.year", "subcategory"]},
+      {"num"=>5, "sort"=>"title|asc", "f"=> @f_app, "facet"=>["creator.name", "date.year", "subcategory"]},
```
Comment (techgique) [inline test/api_bridge_test.rb]: Do you want to re-use `@facet_app` here with `"facet" => @facet_app` similar to `"f" => @f_app`?

File: test/api_bridge_test.rb [hunk 4]
```diff
@@ -122,7 +124,7 @@ def before_setup
     req_opts = ActionController::Parameters.new({
       controller: "items", action: "index", utf8: "✓", q: "water"
     })
-    res_opts = {"num"=>5, "sort"=>"title|asc", "f"=>["collection|letters"], "facet"=>["creator.name", "date.year", "subcategory"], "q"=>"water", "start"=>0}
+    res_opts = {"num"=>5, "sort"=>"title|asc", "f"=>["collection|letters", "keyword|production"], "facet"=>["creator.name", "date.year", "subcategory"], "q"=>"water", "start"=>0}
```
Comment (techgique) [inline test/api_bridge_test.rb]: Re-use `@f_app` and `@facet_app` here too?

File: test/api_bridge_test.rb [hunk 5]
```diff
@@ -131,27 +133,32 @@ def before_setup
...
-    # actually returns ["collection|letters", "collection|whitman", "subcategory|marginalia"]
-    setting = ["collection|whitman", "subcategory|marginalia"]
-    assert_equal setting, @api.prepare_options({ "f" => setting })["f"]
+    # if a filter collides (collection in this case), make sure the newly requested one wins out!
+    assert_equal(
+      ["keyword|production", "collection|whitman", "subcategory|marginalia"],
+      @api.prepare_options({ "f" => ["collection|whitman", "subcategory|marginalia"] })["f"]
+    )
+
+    # check that filter collisions where the request has multiple of the same filter preferences request
```
Comment (techgique) [inline test/api_bridge_test.rb]: I think this and the test above are redundant. They are both just testing whether `keyword|production` is removed, right? Whether the replacement has more than one `keyword` or not doesn't matter to the code being tested from how I read it.

File: app/services/api_bridge/query.rb [hunk 6]
```diff
@@ -113,8 +113,21 @@ def prepare_options(options)
       # combine the incoming params with the app default options
       params_hash = params_to_hash(options)
       opts = @app_options.clone.merge(params_hash) { |key, app, req|
-        if app.is_a?(Array) && req.is_a?(Array)
-          app + req
+        if key == "f"
+          # repeated filters are treated as ANDs and for that
```
Comment (techgique) [inline app/services/api_bridge/query.rb]: This is a little hard to follow, sort of written as run-on sentences.

I suggest something like this
```
# Remove app filter if request contains the same filter.
# Otherwise multiples of a filter are AND-ed in the API query
# and will likely return no results. To preserve any app filters,
# they must be sent along with the request filters.
```

I think you could remove the comment from inside the `reject` block too with this description. I think it covers what we're doing and why pretty well :)

File: test/api_bridge_test.rb (hunk 1 above)
Comment (jduss4) [reply test/api_bridge_test.rb]: this is partly a casualty of me not being able to boot up git gui today :'(

File: test/api_bridge_test.rb (hunk 5 above)
Comment (jduss4) [reply test/api_bridge_test.rb]: I mostly wanted to make sure that both of the new keywords would make it through, partly because originally I had written the code in query.rb differently. I think I will remove the more simplistic test rather than this one.

File: test/api_bridge_test.rb (hunk 1 above)
Comment (techgique) [reply test/api_bridge_test.rb]: :fist_raised: Bad Jess's computer! It's going to have time in the corner later

File: test/api_bridge_test.rb [hunk 7]
```diff
@@ -26,10 +26,14 @@ def before_setup
     assert_equal({ "num" => 50, "facet" => [] }, api.app_options)
 
     # override options
-    facets = ["creator.name", "date.year", "subcategory"]
```
Comment (techgique) [inline test/api_bridge_test.rb]: Good catch. I hadn't noticed the duplicate there!

File: test/api_bridge_test.rb [hunk 8]
```diff
@@ -140,25 +152,36 @@ def before_setup
     assert_equal @f_app, @api.prepare_options({})["f"]
 
     # set a new filter which is not the same as the app wide filter
-    assert_equal ["collection|letters", "keyword|production", "subcategory|memos"], @api.prepare_options({ "f" => ["subcategory|memos"] })["f"]
-
-    # if a filter collides (collection in this case), make sure the newly requested one wins out!
     assert_equal(
-      ["keyword|production", "collection|whitman", "subcategory|marginalia"],
-      @api.prepare_options({ "f" => ["collection|whitman", "subcategory|marginalia"] })["f"]
+      ["collection|letters", "keyword|production", "subcategory|memos"],
```
Comment (techgique) [inline test/api_bridge_test.rb]: I suppose for even more de-duplicating, we could write this as `@f_app + ["subcategory|memos"]`. That mimics the code being tested that concatenates the array contents too. What do you think? :thinking:

Comment (jduss4) [reply test/api_bridge_test.rb]: That's true. I can do either, I don't have a particular preference!

Comment (techgique) [reply test/api_bridge_test.rb]: It'll also simplify rewriting test values if we need to change `@f_app` for future tests. Give it a try?
